File: Python/20040_사이클 게임.py
import sys
sys.setrecursionlimit(10 ** 7)
input = sys.stdin.readline


def find(parent, a):
    if parent[a] == a:
        return a
    
    parent[a] = find(parent, parent[a])
    return parent[a]


def union(parent, a, b):
    a = find(parent, a)
    b = find(parent, b)

    if a == b:
        return True

    if rank[a] > rank[b]:
        parent[b] = a
        rank[a] += 1
    else:
        parent[a] = b
        rank[b] += 1
    
    return False


N, M = map(int, input().split())
rank = [1 for _ in range(N)]
parent = [i for i in range(N)]

result = []
for i in range(1, M+1):
    a, b = map(int, input().split())
    if union(parent, a, b):
        result.append(i)

if result:
    print(result[0])
else:
    print(0)

File: C++/20040_사이클 게임.cpp
#include <iostream>
#include <vector>

using namespace std;

int N, M;
int parent[500001];
int _rank[500001];
vector<int> result;

int find(int a) {
    if (parent[a] == a)
        return a;

    return parent[a] = find(parent[a]);
}

bool vnion(int a, int b) {
    a = find(a);
    b = find(b);

    if (a == b)
        return true;

    if (_rank[a] > _rank[b]) {
        parent[b] = a;
        _rank[a]++;
    } else {
        parent[a] = b;
        _rank[b]++;
    }

    return false;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    cin >> N >> M;
    for (int i = 0; i < N; i++)
        parent[i] = i;

    int a, b;
    for (int i = 1; i <= M; i++) {
        cin >> a >> b;
        if (vnion(a, b))
            result.push_back(i);
    }

    cout << (result.empty() ? 0 : result.front());

    return 0;
}